pkg/clusteragent/admission/mutate/autoinstrumentation/annotation_extractor.go

```diff
@@ -0,0 +1,46 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build kubeapiserver
+
+package autoinstrumentation
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// annotationExtractor extracts and transforms a kubernetes objects annotation
+// per a callback.
+type annotationExtractor[T any] struct {
+	key string
+	do  func(string) (T, error)
+}
+
+// forKey creates a new annotationExtractor for the given key, copying
+// the callback function.
+func (e annotationExtractor[T]) forKey(key string) annotationExtractor[T] {
+	return annotationExtractor[T]{
+		key: key,
+		do:  e.do,
+	}
+}
+
+// extract extracts annotation data from the kubernetes Object.
+func (e annotationExtractor[T]) extract(o metav1.Object) (T, bool, error) {
```

Comment:
Question: 

why do we need bool and error returns here?

isn't it sufficient to just return (T, error) ? 

if the annotation key doesn't exist, we return `NotFound` error.

WDYT?

Reply:
If we use a sentinel error, and errors.Is(), that could work.

Reply:
Updated with the sentinel error and not found check